tests/rptest/tests/partition_balancer_test.py

```diff
@@ -510,13 +505,35 @@ def func(s):
             # Wait until the balancer manages to move partitions from the killed node.
             # This will make other 4 nodes go over the disk usage limit and the balancer
             # will stall because there won't be any other node to move partitions to.
-            self.wait_until_status(create_waiter("stalled"))
+
+            # waiter that prints current node disk usage while it waits.
+            def is_stalled(s):
+                print_disk_usage_per_node()
+                return s["status"] == "stalled"
+
+            self.wait_until_status(is_stalled)
             self.check_no_replicas_on_node(ns.cur_failure.node)
 
             # bring failed node up
             ns.make_available()
 
-        self.wait_until_status(create_waiter("ready"))
+        # Wait until the balancer is done and stable for a few ticks in a row
+        ready_appeared_at = None
+
+        def is_ready_and_stable(s):
+            print_disk_usage_per_node()
+            nonlocal ready_appeared_at
```

Comment:
our wrapper `wait_until_result` is able to return state from the predicate

EDIT: oh, i see with self.wait_until_status below there would need to be multiple levels of wrappers, so it is probably more trouble than it's worth

Reply:
I may have misunderstood your comment, but this `ready_appeared_at` variable is used not to return a value but to save some state between invocations - we remember the time "ready" status appeared and used it to check that it was stable for a few predicate invocations.

Reply:
oh i was mistaken when i saw nonlocal about why it was being used. ignore :)